qemu/tests/single_driver_update.py

```diff
@@ -0,0 +1,170 @@
+import re
+import logging
+from autotest.client.shared import error
+from virttest import utils_misc
+
+
+@error.context_aware
+def run(test, params, env):
+    """
+    This Test is mainly used as subtests
+    1) Boot up VM
+    2) Find the driver from the repo
+    3) record the driver version
+    4) Copy it to C:\ (If there are two copy the newest one)
+    5) Uninstall the driver/downgrade the driver
+    6) Reinstall the driver copied to C:\\tmp
+    7) Check the driver version whether same as original
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def get_installation_logs(session):
+        tmp_debug = "%systemroot%\\DPINST.log,\
+                     %systemroot%\\inf\\setupapi.dev.log"
+        driver_logs = params.get("driver_debug_file", tmp_debug).split(',')
+        error.context("Get Driver installation log", logging.info)
+        for _file in driver_logs:
+            s, o = session.cmd_status_output("dir %s" % _file)
+            if not s:
+                output = session.cmd("type %s" % _file)
+                _file_host = _file.split("\\")[-1].strip()
+                _debug_log = open("%s/%s" % (test.resultsdir, _file_host), "w")
+                _debug_log.write("%s\n" % output)
+                _debug_log.close()
+
+    def backup_latest_folder(session, folders):
+        driver_repo_path = params.get("driver_repo_path")
+        tmp_folder = params.get("tmp_folder", 'C:\tmp')
+        status, output = session.cmd_status_output("dir /B %s" % tmp_folder)
+        if status:
+            session.cmd("mkdir %s" % tmp_folder)
+        latest_folder = ""
+        latest_version = 0
+        if len(folders) == 0:
+            get_installation_logs(session)
+            raise error.TestError("No folder found, \
+                                  Pls check whether the driver is installed")
+        else:
+            for folder in folders:
+                cli = "type %s\\%s\\*.inf" % (driver_repo_path, folder)
+                output_full = session.cmd(cli)
+                key = "DriverVer.*\S"
+                output = re.findall(key, output_full, re.M)[-1]
+                if int(output.split('.')[-1]) >= latest_version:
+                    latest_folder = folder
+                    latest_version = int(output.split('.')[-1])
+        logging.info("Latest installed version is %s"
+                     % latest_version)
+        session.cmd("xcopy %s\%s %s\%s /S /Y /I"
+                    % (driver_repo_path,
+                       latest_folder,
+                       tmp_folder,
+                       latest_folder))
+        return latest_folder
+
+    def driver_folder_find(session):
+        driver_name = params.get("driver_name")
+        driver_repo_path = params.get("driver_repo_path")
+        search_cmd = "dir /B  %s\%s.inf*" % (driver_repo_path, driver_name)
```

Comment:
Suggest to add step description here
 error.context("Check driver folders in %s" % driver_repo_path)